File: medium/container-with-most-water.go
package main

import (
	"fmt"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

/*
https://leetcode.com/problems/container-with-most-water/
某个vertical能产生的最大面积:找到比自己短的vertical的位置
*/
func maxArea(height []int) int {
	start := 0
	end := len(height) - 1
	startHeight := height[start]
	endHeight := height[end]
	area := min(startHeight, endHeight) * (end - start)
	var moveStart bool
	for start < end {
		if startHeight < endHeight {
			moveStart = true
		} else {
			moveStart = false
		}
		if moveStart {
			start++
			if height[start] > startHeight {
				startHeight = height[start]
				area = max(area, min(startHeight, endHeight)*(end-start))
			}
		} else {
			end--
			if height[end] > endHeight {
				endHeight = height[end]
				area = max(area, min(startHeight, endHeight)*(end-start))
			}
		}
	}
	return area
}

func main() {
	fmt.Println(maxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}) == 49)
	fmt.Println(maxArea([]int{4, 3, 2, 1, 4}) == 16)
	fmt.Println(maxArea([]int{1, 2, 1}) == 2)
}

File: medium/container-with-most-water.py
# coding=utf8

"""
Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

Note: You may not slant the container and n is at least 2.
"""
from typing import List


class Solution(object):
    def maxArea(self, height: List[int]) -> int:
        """
        :type height: List[int]
        :rtype: int
        """
        i = 0
        j = len(height) - 1
        maxAero = 0
        while i < j:
            minHeight = min(height[i], height[j])
            maxAero = max(maxAero, (j - i) * minHeight)

            if height[i] > height[j]:
                j -= 1
            else:
                i += 1
        return maxAero
